File: lib/search_index.rb
class SearchIndex
  def self.generate
    snippets =
      Snippet.
        listed.
        preload(:language, :collections).
        sort_by(&:cid)
    collections =
      Collection.
        listed.
        preload(:collection_snippets, :parent, :children, snippets: [:language]).
        sort_by(&:cid)

    search_index =
      { search_index: SearchResultSerializer.serialize_array(snippets + collections) }.
        transform_keys! do |key|
          key.to_s.camelize(:lower)
        end

    File.write(
      'public/search-data.json',
      JSON.pretty_generate(search_index)
    )
  end
end

File: src/lib/searchIndex.js
import fs from 'fs-extra/esm';
import Collection from '#src/models/collection.js';
import Snippet from '#src/models/snippet.js';
import SearchResultSerializer from '#src/serializers/searchResultSerializer.js';

export default class SearchIndex {
  static generate() {
    const snippets = Snippet.scope('published', 'listed', 'byId');
    const collections = Collection.scope('listed', 'byId');

    const searchIndex = {
      searchIndex: SearchResultSerializer.serializeArray(
        snippets.concat(collections)
      ),
    };

    // Write to file
    fs.writeJson(
      'public/search-data.json',
      searchIndex,
      { spaces: 0 },
      () => {}
    );
  }
}
